lib/http-fetch.ts

```diff
@@ -0,0 +1,189 @@
+    if (this.responseParser) return this.responseParser(res);
+    else return await res.json();
+  }
+
+  public async put<T>(
+    url: string,
+    body?: any,
+    config?: Partial<FetchRequestConfig>,
+  ): Promise<T> {
+    const requestUrl = new URL(url, this.baseURL);
+    const response = await fetch(requestUrl, {
+      method: "PUT",
+      headers: {
+        "Content-Type": "application/json",
+        ...this.defaultHeaders,
+        ...(config && config.headers),
+      },
+      body: JSON.stringify(body),
+    });
+    this.checkResponseStatus(response);
+    return this.responseParse(response);
+  }
+
+  public async postForm<T>(url: string, body?: any): Promise<T> {
+    const requestUrl = new URL(url, this.baseURL);
+    const response = await fetch(requestUrl, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded",
+        ...this.defaultHeaders,
+      },
+      body: qs.stringify(body),
+    });
+    this.checkResponseStatus(response);
+    return response.json();
+  }
+
+  public async postFormMultipart<T>(url: string, form: FormData): Promise<T> {
+    const requestUrl = new URL(url, this.baseURL);
+    const response = await fetch(requestUrl, {
+      method: "POST",
+      headers: {
+        ...this.defaultHeaders,
+      },
+      body: form,
+    });
+    this.checkResponseStatus(response);
+    return response.json();
+  }
+
+  public async putFormMultipart<T>(
+    url: string,
+    form: FormData,
+    config?: Partial<FetchRequestConfig>,
+  ): Promise<T> {
+    const requestUrl = new URL(url, this.baseURL);
+    const response = await fetch(requestUrl, {
+      method: "PUT",
+      headers: {
+        ...this.defaultHeaders,
+        ...(config && (config.headers ? config.headers : {})),
+      },
+      body: form,
+    });
+    this.checkResponseStatus(response);
+    return response.json();
+  }
+  public async postBinaryContent<T>(url: string, body: Blob): Promise<T> {
+    const requestUrl = new URL(url, this.baseURL);
+    const response = await fetch(requestUrl, {
+      method: "POST",
+      headers: {
+        "Content-Type": body.type,
+        ...this.defaultHeaders,
+      },
+      body: body,
+    });
+    this.checkResponseStatus(response);
+    return response.json();
+  }
+
+  public async delete<T>(url: string, params?: any): Promise<T> {
+    const requestUrl = new URL(url, this.baseURL);
+    if (params) {
+      requestUrl.search = new URLSearchParams(params).toString();
+    }
+    const response = await fetch(requestUrl, {
+      method: "DELETE",
+      headers: {
+        ...this.defaultHeaders,
+      },
+    });
+    this.checkResponseStatus(response);
+    return response.json();
+  }
+
+  private checkResponseStatus(response: Response) {
+    if (!response.ok) {
+      throw new HTTPError(
+        "HTTP request failed",
+        response.status,
+        response.statusText,
+        undefined,
+      );
```

Comment:
can we make this include actual error message(error body) and headers? (or this is resolved in 2nd step?)

Reply:
Good catch.

Ah, there's some option at this point.
Currently, fourth argument of the HTTPError is an exception type.

So, current implementation set this to `axios`'s exception object. But there's no equivalent for **fetch**.

I guess there's 3 options at here.

1. Create new error type and pass it to the fourth argument of the HTTPError.
2. Create new error type, that replaces HTTPError entirely. It may break backward compatibility. But it's very clean interface.
3. Add fifth argument to the HTTPError, that receives response content itself.

Which is your preferred choice?

Reply:
Thank you for providing the options! No matter which option we choose, compatibility will be broken the moment we start using fetch. line-bot-sdk-nodejs users will have to make some changes in their application. Therefore, I prefer option 2.

Retrieving the existing error body is very painful. I think option 1 is good as a workaround, but it doesn't solve the following:

| Property   | Current Access         | New Access            |
|------------|-----------------------|-----------------------|
| status code| `e.originalError.status` | `e.newError.status` |
| errorBody  | `e.originalError.data(?)` | `e.newError.data(?)` |
| headers    | No way to get them    | `e.newError.headers` |

Option 3 might be good, but I think it's confusing for clients using axios and those using fetch to use the same exception.